src/global/project/mod.rs

```diff
@@ -80,31 +105,257 @@ impl Project {
         }
+
         Self::from_path(&manifest_path)
     }
 
+    async fn from_existing_installation(
+        manifest_path: &Path,
+        bin_dir: &BinDir,
+        env_root: &EnvRoot,
+    ) -> miette::Result<Self> {
+        let exposed_binaries: Vec<ExposedData> = bin_dir
+            .files()
+            .await?
+            .into_iter()
+            .filter_map(|path| match is_text(&path) {
+                Ok(true) => Some(Ok(path)), // Success and is text, continue with path
+                Ok(false) => None,          // Success and isn't text, filter out
+                Err(e) => Some(Err(e)),     // Failure, continue with error
+            })
+            .map(|result| result.and_then(Self::exposed_data_from_binary_path))
+            .collect::<miette::Result<_>>()?;
+
+        let parsed_manifest = ParsedManifest::from(exposed_binaries);
+        let toml = toml_edit::ser::to_string(&parsed_manifest).into_diagnostic()?;
+        tokio::fs::write(&manifest_path, &toml)
+            .await
+            .into_diagnostic()?;
+        Self::from_str(manifest_path, &toml)
+    }
+
+    fn exposed_data_from_binary_path(path: PathBuf) -> miette::Result<ExposedData> {
+        let exposed = path
+            .file_stem()
+            .and_then(OsStr::to_str)
+            .ok_or_else(|| miette::miette!("Could not get file stem of {}", path.display()))
+            .and_then(ExposedKey::from_str)?;
+        let binary_path = Self::extract_bin_from_script(&path)?;
+
+        let binary = binary_path
+            .file_stem()
+            .and_then(OsStr::to_str)
+            .map(String::from)
+            .ok_or_else(|| miette::miette!("Could not get file stem of {}", path.display()))?;
+        let env_path = binary_path
+            .parent()
+            .ok_or_else(|| {
+                miette::miette!("binary_path '{}' has no parent", binary_path.display())
+            })?
+            .parent()
+            .ok_or_else(|| {
+                miette::miette!(
+                    "binary_path's parent '{}' has no parent",
+                    binary_path.display()
+                )
+            })?;
+        let env = env_path
+            .file_name()
+            .and_then(OsStr::to_str)
+            .ok_or_else(|| {
+                miette::miette!(
+                    "binary_path's grandparent '{}' has no file name",
+                    binary_path.display()
+                )
+            })
+            .and_then(|env| EnvironmentName::from_str(env).into_diagnostic())?;
+
+        let conda_meta = env_path.join("conda-meta");
+
+        let (platform, channel, package) = Self::package_from_conda_meta(&conda_meta, &binary)?;
+
+        Ok(ExposedData {
+            env,
+            platform,
+            channel,
+            package,
+            binary,
+            exposed,
+        })
+    }
+
+    fn package_from_conda_meta(
```

Comment:
I think we need to make this functions async as well because we are working with filesystem , which are blocking operations, and by using them from async upstream I think we can block event loop here

Reply:
Good point, changed!